# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/basetypes/Accessible.cpp
# Commit: ac04b727a98b
# Full Hash: ac04b727a98bb1d4e0644665b59ed4da3e9e3858
# Author: Eitan Isaacson <eitan@monotonous.org>
# Date: 2022-01-05 09:33:53
# Regressor Bug: 1737944
# File Overlap Count: 2
# Description:
#   Bug 1737944 - P8: Move GetGroupPosition to Accessible. r=Jamie
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D134211
# ==============================================================================

diff -r 5d76535b47fc -r ac04b727a98b accessible/basetypes/Accessible.cpp
--- a/accessible/basetypes/Accessible.cpp	Tue Jan 04 21:01:37 2022 +0000
+++ b/accessible/basetypes/Accessible.cpp	Tue Jan 04 21:01:37 2022 +0000
@@ -80,6 +80,35 @@
   return hyperText ? hyperText->GetChildOffset(this) : 0;
 }
 
+GroupPos Accessible::GroupPosition() {
+  GroupPos groupPos;
+
+  // Get group position from ARIA attributes.
+  ARIAGroupPosition(&groupPos.level, &groupPos.setSize, &groupPos.posInSet);
+
+  // If ARIA is missed and the accessible is visible then calculate group
+  // position from hierarchy.
+  if (State() & states::INVISIBLE) return groupPos;
+
+  // Calculate group level if ARIA is missed.
+  if (groupPos.level == 0) {
+    groupPos.level = GetLevel(false);
+  }
+
+  // Calculate position in group and group size if ARIA is missed.
+  if (groupPos.posInSet == 0 || groupPos.setSize == 0) {
+    int32_t posInSet = 0, setSize = 0;
+    GetPositionAndSetSize(&posInSet, &setSize);
+    if (posInSet != 0 && setSize != 0) {
+      if (groupPos.posInSet == 0) groupPos.posInSet = posInSet;
+
+      if (groupPos.setSize == 0) groupPos.setSize = setSize;
+    }
+  }
+
+  return groupPos;
+}
+
 int32_t Accessible::GetLevel(bool aFast) const {
   int32_t level = 0;
   if (!Parent()) return level;